torch/distributed/pipelining/PipelineSchedule.py

```diff
@@ -685,6 +697,130 @@ def __init__(
 
         self.n_local_stages = len(stages)
         self.rank = stages[0].group_rank
+        self.group = stages[0].group
+
+        # 1. Create the pipeline_order (all ranks do this calculation)
+        # This will be used to keep track of the current state of the entire pipeline
+        # pipeline_order[rank] = [Action(computation_type, microbatch_index, stage_index), ...]
+        self.pipeline_order: Dict[int, List[Optional[_Action]]] = {}
+        # ========================================================================
+        for rank in range(self.pp_group_size):
+            rank_ops = self._calculate_single_rank_operations(rank)
+            self.pipeline_order[rank] = rank_ops
+
+    def _calculate_single_rank_operations(self, rank) -> List[Optional[_Action]]:
+        def get_rank_warmup_ops(rank):
+            # Warms up operations for last stage
+            warmups_ops_last_stage = (self.n_local_stages - 1) * self.pp_group_size
```

Comment:
if we were to use this logic for the non-looped case (n_local_stages==1) this logic would break.  Is there a more robust formula we could use?

perhaps, computing this relative to the globally last stage first, and then just offsetting it by the local stage_id?

Reply:
I think each schedule can implement its own function for the `_calculate_single_rank_operations`, in loopedBFS this entire function is different (https://github.com/pytorch/pytorch/pull/127796/files#diff-bcea404838585952f926a03a7da41b33f8bdc541a955ef25345a69ab2e2a286eR727). What could be shared is the `step_microbatches` logic.

> perhaps, computing this relative to the globally last stage first, and then just offsetting it by the local stage_id?

I could try, but i think there may be some edge case with this

Reply:
ok. better to refactor later anyway, i guess.